Intermittent failure of test_rx_rate_limiting test

test_rx_rate_limiting is failing intermittently.
Example of output:

```
=================================== FAILURES ===================================
____________________ test_rx_rate_limiting[ubuntu_with_ssh] ____________________
integration_tests/functional/test_rate_limiter.py:147: in test_rx_rate_limiting
    _check_rx_rate_limiting(test_microvm, guest_ips)
integration_tests/functional/test_rate_limiter.py:283: in _check_rx_rate_limiting
    assert (
E   assert 11.539077758789062 < 10
E    +  where 11.539077758789062 = _get_difference(92758000.0, 104857600.0)
------------------------------ Captured log setup ------------------------------
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8375] Command:
cd /firecracker && TARGET_CC=musl-gcc cargo build --release --target x86_64-unknown-linux-musl
[8375] stderr:
    Finished release [optimized] target(s) in 0.03s
------------------------------ Captured log call -------------------------------
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8378] Command:
ip netns add 7f369f93-a01c-4183-b4ec-00dd48078101
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8380] Command:
ln -f /tmp/fctest-efdxs5su/7f369f93-a01c-4183-b4ec-00dd48078101/log_fifo /srv/jailer/firecracker/7f369f93-a01c-4183-b4ec-00dd48078101/root/log_fifo
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8382] Command:
chown 1234:1234 /srv/jailer/firecracker/7f369f93-a01c-4183-b4ec-00dd48078101/root/log_fifo
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8385] Command:
['/tmp/fctest-efdxs5su/newpid_cloner', '/firecracker/build/cargo_target/x86_64-unknown-linux-musl/release/jailer', '--id', '7f369f93-a01c-4183-b4ec-00dd48078101', '--exec-file', '/firecracker/build/cargo_target/x86_64-unknown-linux-musl/release/firecracker', '--node', '0', '--uid', '1234', '--gid', '1234', '--chroot-base-dir', '/srv/jailer', '--netns', '/var/run/netns/7f369f93-a01c-4183-b4ec-00dd48078101', '--daemonize', '--', '--log-path', 'log_fifo', '--level', 'Info']
[8385] stdout:
8386
09:51:19 retry.api: WARNING , retrying in 0.1 seconds...
09:51:19 urllib3.connectionpool: DEBUG http://localhost:None "PUT /machine-config HTTP/1.1" 204 0
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8389] Command:
pmap -xq 8386
[8389] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1024    1024 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000      28      28      28 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bd6d000     132       8       8 rw---   [ stack ]
00007ffc3bd6d000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:19 commands: DEBUG 
[8392] Command:
ln -f /tmp/fctest-efdxs5su/7f369f93-a01c-4183-b4ec-00dd48078101/kernel/vmlinux.bin /srv/jailer/firecracker/7f369f93-a01c-4183-b4ec-00dd48078101/root/vmlinux.bin
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8394] Command:
chown 1234:1234 /srv/jailer/firecracker/7f369f93-a01c-4183-b4ec-00dd48078101/root/vmlinux.bin
09:51:19 urllib3.connectionpool: DEBUG http://localhost:None "PUT /boot-source HTTP/1.1" 204 0
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8396] Command:
ln -f /tmp/fctest-efdxs5su/7f369f93-a01c-4183-b4ec-00dd48078101/fsfiles/xenial.rootfs.ext4 /srv/jailer/firecracker/7f369f93-a01c-4183-b4ec-00dd48078101/root/xenial.rootfs.ext4
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8398] Command:
chown 1234:1234 /srv/jailer/firecracker/7f369f93-a01c-4183-b4ec-00dd48078101/root/xenial.rootfs.ext4
09:51:19 urllib3.connectionpool: DEBUG http://localhost:None "PUT /drives/rootfs HTTP/1.1" 204 0
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8400] Command:
ip tuntap add mode tap name 7f369f93tap1
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8402] Command:
ip link set 7f369f93tap1 netns 7f369f93-a01c-4183-b4ec-00dd48078101
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8404] Command:
ip netns exec 7f369f93-a01c-4183-b4ec-00dd48078101 ifconfig 7f369f93tap1 192.168.0.21/30 up
09:51:19 urllib3.connectionpool: DEBUG http://localhost:None "PUT /network-interfaces/1 HTTP/1.1" 204 0
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:19 commands: DEBUG 
[8406] Command:
ip tuntap add mode tap name 7f369f93tap2
09:51:19 asyncio: DEBUG Using selector: EpollSelector
09:51:20 commands: DEBUG 
[8408] Command:
ip link set 7f369f93tap2 netns 7f369f93-a01c-4183-b4ec-00dd48078101
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:20 commands: DEBUG 
[8410] Command:
ip netns exec 7f369f93-a01c-4183-b4ec-00dd48078101 ifconfig 7f369f93tap2 192.168.0.25/30 up
09:51:20 urllib3.connectionpool: DEBUG http://localhost:None "PUT /network-interfaces/2 HTTP/1.1" 204 0
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:20 commands: DEBUG 
[8412] Command:
ip tuntap add mode tap name 7f369f93tap3
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:20 commands: DEBUG 
[8414] Command:
ip link set 7f369f93tap3 netns 7f369f93-a01c-4183-b4ec-00dd48078101
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:20 commands: DEBUG 
[8416] Command:
ip netns exec 7f369f93-a01c-4183-b4ec-00dd48078101 ifconfig 7f369f93tap3 192.168.0.29/30 up
09:51:20 urllib3.connectionpool: DEBUG http://localhost:None "PUT /network-interfaces/3 HTTP/1.1" 204 0
09:51:20 urllib3.connectionpool: DEBUG http://localhost:None "PUT /actions HTTP/1.1" 204 0
09:51:20 nsenter: DEBUG Entering net namespace /var/run/netns/7f369f93-a01c-4183-b4ec-00dd48078101
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:20 commands: DEBUG 
[8421] Command:
pmap -xq 8386
[8421] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   73020   73020 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:20 commands: DEBUG 
[8420] Command:
['ssh', '-q', '-o', 'ConnectTimeout=1', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-i', '/tmp/fctest-efdxs5su/7f369f93-a01c-4183-b4ec-00dd48078101/fsfiles/xenial.rootfs.id_rsa', 'root@192.168.0.22', 'true']
09:51:20 nsenter: DEBUG Leaving net namespace /var/run/netns/7f369f93-a01c-4183-b4ec-00dd48078101
09:51:20 nsenter: DEBUG Entering net namespace /var/run/netns/7f369f93-a01c-4183-b4ec-00dd48078101
09:51:20 asyncio: DEBUG Using selector: EpollSelector
09:51:21 commands: DEBUG 
[8423] Command:
['ssh', '-q', '-o', 'ConnectTimeout=1', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-i', '/tmp/fctest-efdxs5su/7f369f93-a01c-4183-b4ec-00dd48078101/fsfiles/xenial.rootfs.id_rsa', 'root@192.168.0.22', 'iperf3 -sD -f KBytes\n']
09:51:21 nsenter: DEBUG Leaving net namespace /var/run/netns/7f369f93-a01c-4183-b4ec-00dd48078101
09:51:21 asyncio: DEBUG Using selector: EpollSelector
09:51:23 asyncio: DEBUG Using selector: EpollSelector
09:51:23 commands: DEBUG 
[8424] Command:
pmap -xq 8386
[8424] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   73104   73104 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:24 asyncio: DEBUG Using selector: EpollSelector
09:51:24 commands: DEBUG 
[8428] Command:
pmap -xq 8386
[8428] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   73508   73508 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:25 asyncio: DEBUG Using selector: EpollSelector
09:51:25 commands: DEBUG 
[8430] Command:
pmap -xq 8386
[8430] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   73508   73508 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:26 asyncio: DEBUG Using selector: EpollSelector
09:51:26 commands: DEBUG 
[8432] Command:
pmap -xq 8386
[8432] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   73516   73516 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:26 commands: DEBUG 
[8426] Command:
ip netns exec 7f369f93-a01c-4183-b4ec-00dd48078101  iperf3 -c 192.168.0.22 -t3 -f KBytes
[8426] stdout:
Connecting to host 192.168.0.22, port 5201
[  4] local 192.168.0.21 port 43634 connected to 192.168.0.22 port 5201
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[  4]   0.00-1.00   sec  3.24 GBytes  3397961 KBytes/sec    4    935 KBytes       
[  4]   1.00-2.00   sec  3.27 GBytes  3424657 KBytes/sec    0    935 KBytes       
[  4]   2.00-3.00   sec  3.30 GBytes  3462204 KBytes/sec    0    936 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-3.00   sec  9.81 GBytes  3428274 KBytes/sec    4             sender
[  4]   0.00-3.00   sec  9.81 GBytes  3427357 KBytes/sec                  receiver

iperf Done.

09:51:26 asyncio: DEBUG Using selector: EpollSelector
09:51:27 asyncio: DEBUG Using selector: EpollSelector
09:51:27 commands: DEBUG 
[8436] Command:
pmap -xq 8386
[8436] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   74452   74452 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:28 asyncio: DEBUG Using selector: EpollSelector
09:51:28 commands: DEBUG 
[8438] Command:
pmap -xq 8386
[8438] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   74464   74464 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:29 asyncio: DEBUG Using selector: EpollSelector
09:51:29 commands: DEBUG 
[8440] Command:
pmap -xq 8386
[8440] stdout:
8386:   /firecracker --id 7f369f93-a01c-4183-b4ec-00dd48078101 --start-time-us 6947503201 --start-time-cpu-us 207 --log-path log_fifo --level Info
0000000000400000    1360    1216    1216 r-x-- firecracker
0000000000400000       0       0       0 r-x-- firecracker
0000000000753000      56      56      56 rw--- firecracker
0000000000753000       0       0       0 rw--- firecracker
0000000000761000       8       8       8 rw---   [ anon ]
0000000000761000       0       0       0 rw---   [ anon ]
000000000185d000     540     480     480 rw---   [ anon ]
000000000185d000       0       0       0 rw---   [ anon ]
00007fda86d6f000      16       0       0 rw---   [ anon ]
00007fda86d6f000       0       0       0 rw---   [ anon ]
00007fda86d73000       8       0       0 -----   [ anon ]
00007fda86d73000       0       0       0 -----   [ anon ]
00007fda86d75000    2052       8       8 rw---   [ anon ]
00007fda86d75000       0       0       0 rw---   [ anon ]
00007fda86f76000       8       0       0 -----   [ anon ]
00007fda86f76000       0       0       0 -----   [ anon ]
00007fda86f78000    2052       8       8 rw---   [ anon ]
00007fda86f78000       0       0       0 rw---   [ anon ]
00007fda87179000      12       4       4 rw-s-   [ anon ]
00007fda87179000       0       0       0 rw-s-   [ anon ]
00007fda8717c000      12       4       4 rw-s-   [ anon ]
00007fda8717c000       0       0       0 rw-s-   [ anon ]
00007fda8717f000       4       4       4 rw-s- zero (deleted)
00007fda8717f000       0       0       0 rw-s- zero (deleted)
00007fda87180000       4       4       4 rw-s- zero (deleted)
00007fda87180000       0       0       0 rw-s- zero (deleted)
00007fda87181000      12      12      12 rw-s- zero (deleted)
00007fda87181000       0       0       0 rw-s- zero (deleted)
00007fda87184000  262144   74464   74464 rw---   [ anon ]
00007fda87184000       0       0       0 rw---   [ anon ]
00007fda97184000       8       0       0 rw---   [ anon ]
00007fda97184000       0       0       0 rw---   [ anon ]
00007fda97186000       8       0       0 -----   [ anon ]
00007fda97186000       0       0       0 -----   [ anon ]
00007fda97188000    2064      12      12 rw---   [ anon ]
00007fda97188000       0       0       0 rw---   [ anon ]
00007ffc3bccb000     780     780     780 rw---   [ stack ]
00007ffc3bccb000       0       0       0 rw---   [ stack ]
00007ffc3bdb0000      12       0       0 r----   [ anon ]
00007ffc3bdb0000       0       0       0 r----   [ anon ]
00007ffc3bdb3000       8       4       0 r-x--   [ anon ]
00007ffc3bdb3000       0       0       0 r-x--   [ anon ]
ffffffffff600000       4       0       0 r-x--   [ anon ]
ffffffffff600000       0       0       0 r-x--   [ anon ]

09:51:29 commands: DEBUG 
[8434] Command:
ip netns exec 7f369f93-a01c-4183-b4ec-00dd48078101  iperf3 -c 192.168.0.26 -t3 -f KBytes
[8434] stdout:
Connecting to host 192.168.0.26, port 5201
[  4] local 192.168.0.25 port 54290 connected to 192.168.0.26 port 5201
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[  4]   0.00-1.00   sec  76.3 MBytes  78088 KBytes/sec  101    182 KBytes       
[  4]   1.00-2.00   sec   103 MBytes  105437 KBytes/sec    2    954 KBytes       
[  4]   2.00-3.00   sec  92.5 MBytes  94750 KBytes/sec    2    954 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-3.00   sec   272 MBytes  92758 KBytes/sec  105             sender
[  4]   0.00-3.00   sec   270 MBytes  91994 KBytes/sec                  receiver

iperf Done.
---------------------------- Captured log teardown -----------------------------
09:51:29 asyncio: DEBUG Using selector: EpollSelector
09:51:29 commands: DEBUG 
[8442] Command:
kill -9 8386
============= 1 failed, 71 passed, 1 skipped in 418.53s (0:06:58) ============== 
```

But, if it helps at all, failure rates for me are: 

4.9 - 60%
5.7 - 10%
4.14 - 10% 

So maybe it's fairer to say that 4.9 seems particularly bad, for me at least, rather than 5.7 fixes it.....

```
strings vmlinux.bin | grep '^Linux version'
Linux version 4.9.0+ (seriorga@u636c053592cf59.ant.amazon.com) (gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1) ) #96 SMP Tue Mar 3 08:56:10 EET 2020


for x in {1..10}; do ./tools/devtool test -- integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting |grep ===; done
============================= test session starts ==============================
============================== 1 passed in 31.70s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 16.47s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 16.21s ==============================
============================= test session starts ==============================
============================== 1 passed in 31.71s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 22.00s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 16.25s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 16.31s ==============================
============================= test session starts ==============================
============================== 1 passed in 32.39s ==============================
============================= test session starts ==============================
============================== 1 passed in 33.42s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 16.69s ==============================

strings vmlinux.bin | grep '^Linux version'
Linux version 5.7.0-rc4 (xxxx@xxxxxx) (gcc version 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC), GNU ld version 2.32-31.fc31) #2 SMP Fri May 8 14:49:02 BST 2020
for x in {1..10}; do ./tools/devtool test -- integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting |grep ===; done
============================= test session starts ==============================
============================== 1 passed in 30.23s ==============================
============================= test session starts ==============================
============================== 1 passed in 30.00s ==============================
============================= test session starts ==============================
============================== 1 passed in 29.92s ==============================
============================= test session starts ==============================
============================== 1 passed in 30.14s ==============================
============================= test session starts ==============================
============================== 1 passed in 31.22s ==============================
============================= test session starts ==============================
============================== 1 passed in 31.39s ==============================
============================= test session starts ==============================
============================== 1 passed in 32.58s ==============================
============================= test session starts ==============================
============================== 1 passed in 33.53s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 15.90s ==============================
============================= test session starts ==============================
============================== 1 passed in 29.90s ==============================

strings vmlinux.bin | grep '^Linux version'
Linux version 4.14.55-84.37.amzn2.x86_64 (mockbuild@ip-10-0-1-79) (gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC)) #1 SMP Wed Jul 25 18:47:15 UTC 2018

for x in {1..10}; do ./tools/devtool test -- integration_tests/functional/test_rate_limiter.py::test_rx_rate_limiting |grep ===; done
============================= test session starts ==============================
============================== 1 passed in 29.99s ==============================
============================= test session starts ==============================
============================== 1 passed in 31.55s ==============================
============================= test session starts ==============================
============================== 1 passed in 30.43s ==============================
============================= test session starts ==============================
============================== 1 passed in 30.85s ==============================
============================= test session starts ==============================
=================================== FAILURES ===================================
============================== 1 failed in 15.03s ==============================
============================= test session starts ==============================
============================== 1 passed in 30.67s ==============================
============================= test session starts ==============================
============================== 1 passed in 30.47s ==============================
============================= test session starts ==============================
============================== 1 passed in 31.70s ==============================
============================= test session starts ==============================
============================== 1 passed in 31.84s ==============================
============================= test session starts ==============================
============================== 1 passed in 29.98s ==============================
```

--- a/tests/integration_tests/functional/test_rate_limiter.py
+++ b/tests/integration_tests/functional/test_rate_limiter.py
@@ -10,20 +10,23 @@ import host_tools.network as net_tools  # pylint: disable=import-error
 IPERF_BINARY = 'iperf3'
 
 # Interval used by iperf to get maximum bandwidth
-IPERF_TRANSMIT_TIME = 3
+IPERF_TRANSMIT_TIME = 4
 
 # Interval to omit statistics collection at the beginning of the test, to skip
 # past the TCP slow-start period
-IPERF_OMIT_TIME = 2
+IPERF_OMIT_TIME = 1
+
+# Use a fixed-size TCP window so we get constant flow
+IPERF_TCP_WINDOW = '1000K'
 
 # The rate limiting value
 RATE_LIMIT_BYTES = 10485760
 
 # The initial token bucket size
-BURST_SIZE = 1048576000
+BURST_SIZE = 104857600
 
 # The refill time for the token bucket
-RATE_LIMIT_REFILL_TIME = 100
+REFILL_TIME_MS = 100
 
 # Deltas that are accepted between expected values and achieved
 # values throughout the tests
@@ -58,7 +61,7 @@ def test_tx_rate_limiting(test_microvm_with_ssh, network_config):
     tx_rate_limiter_no_burst = {
         'bandwidth': {
             'size': RATE_LIMIT_BYTES,
-            'refill_time': RATE_LIMIT_REFILL_TIME
+            'refill_time': REFILL_TIME_MS
         }
     }
     _tap2, host_ip, guest_ip = test_microvm.ssh_network_config(
@@ -74,7 +77,7 @@ def test_tx_rate_limiting(test_microvm_with_ssh, network_config):
         'bandwidth': {
             'size': RATE_LIMIT_BYTES,
             'one_time_burst': BURST_SIZE,
-            'refill_time': RATE_LIMIT_REFILL_TIME
+            'refill_time': REFILL_TIME_MS
         }
     }
     _tap3, host_ip, guest_ip = test_microvm.ssh_network_config(
@@ -118,7 +121,7 @@ def test_rx_rate_limiting(test_microvm_with_ssh, network_config):
     rx_rate_limiter_no_burst = {
         'bandwidth': {
             'size': RATE_LIMIT_BYTES,
-            'refill_time': RATE_LIMIT_REFILL_TIME
+            'refill_time': REFILL_TIME_MS
         }
     }
     _tap2, host_ip, guest_ip = test_microvm.ssh_network_config(
@@ -134,7 +137,7 @@ def test_rx_rate_limiting(test_microvm_with_ssh, network_config):
         'bandwidth': {
             'size': RATE_LIMIT_BYTES,
             'one_time_burst': BURST_SIZE,
-            'refill_time': RATE_LIMIT_REFILL_TIME
+            'refill_time': REFILL_TIME_MS
         }
     }
     _tap3, host_ip, guest_ip = test_microvm.ssh_network_config(
@@ -158,95 +161,51 @@ def _check_tx_rate_limiting(test_microvm, guest_ips, host_ips):
     _start_local_iperf(test_microvm.jailer.netns_cmd_prefix())
 
     # First step: get the transfer rate when no rate limiting is enabled.
-    # We are receiving the result in KBytes from iperf; 1000 converts to Bytes.
-    iperf_cmd = '{} -c {} -t{} -f KBytes -O {}'.format(
+    # We are receiving the result in KBytes from iperf.
+    iperf_cmd = '{} -c {} -t {} -f KBytes -O {} -w {} -N'.format(
         IPERF_BINARY,
         host_ips[0],
         IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME
+        IPERF_OMIT_TIME,
+        IPERF_TCP_WINDOW
     )
-
+    print("Run guest TX iperf with no rate-limit")
     iperf_out = _run_iperf_on_guest(test_microvm, iperf_cmd, guest_ips[0])
-    iperf_out = _process_iperf_output(iperf_out)[1]
-
-    rate_no_limit_bytes = 1000 * float(iperf_out)
-
-    # Second step: get the number of bytes when rate limiting is on.
+    _, rate_no_limit_kbps = _process_iperf_output(iperf_out)
+    print("TX rate_no_limit_kbps: {}".format(rate_no_limit_kbps))
 
     # Calculate the number of bytes that are expected to be sent
     # in each second once the rate limiting is enabled.
-    rate_limit_bps = 1000 * RATE_LIMIT_BYTES / float(RATE_LIMIT_REFILL_TIME)
+    expected_kbps = int(RATE_LIMIT_BYTES / (REFILL_TIME_MS / 1000.0) / 1024)
+    print("Configured TX expected_kbps: {}".format(expected_kbps))
 
-    # Use iperf for some number of seconds to get the number of bytes it sent
-    # with rate limiting on.
-    iperf_cmd = '{} -c {} -t{} -f KBytes'.format(
-        IPERF_BINARY,
-        host_ips[1],
-        IPERF_TRANSMIT_TIME
-    )
-    iperf_out = _run_iperf_on_guest(
-        test_microvm, iperf_cmd, guest_ips[1]
-    )
-    iperf_out = _process_iperf_output(iperf_out)[1]
+    # Sanity check that bandwidth with no rate limiting is at least double
+    # than the one expected when rate limiting is in place.
+    assert _get_percentage_difference(rate_no_limit_kbps, expected_kbps) > 100
 
-    rate_limit_bytes_achieved = 1000 * float(iperf_out)
-    rate_limit_bytes_expected = rate_limit_bps
-
-    # Assert on the bytes expected and achieved with rate limiting on; we are
-    # expecting a difference no bigger than MAX_RATE_LIMIT_BYTES_DIFF
-    assert (
-            _get_difference(
-                rate_limit_bytes_achieved,
-                rate_limit_bytes_expected
-            )
-            < MAX_BYTES_DIFF_PERCENTAGE
-    )
+    # Second step: check bandwith when rate limiting is on.
+    _check_tx_bandwidth(test_microvm, guest_ips[1], host_ips[1], expected_kbps)
 
     # Third step: get the number of bytes when rate limiting is on and there is
     # an initial burst size from where to consume.
 
+    print("Run guest TX iperf with exact burst size")
     # Use iperf to obtain the bandwidth when there is burst to consume from.
-    iperf_cmd = '{} -c {} -n{} -f KBytes'.format(
+    iperf_cmd = '{} -c {} -n {} -f KBytes -w {} -N'.format(
         IPERF_BINARY,
         host_ips[2],
-        BURST_SIZE
+        BURST_SIZE,
+        IPERF_TCP_WINDOW
     )
     iperf_out = _run_iperf_on_guest(test_microvm, iperf_cmd, guest_ips[2])
-    # iperf will give variable number of output lines depending on how much
-    # time it took to send the amount specified.
-    burst_bw_first_time_achieved = _process_iperf_output(iperf_out)[1]
+    print(iperf_out)
+    _, burst_kbps = _process_iperf_output(iperf_out)
+    print("TX burst_kbps: {}".format(burst_kbps))
+    # Test that the burst bandwidth is at least as two times the rate limit.
+    assert _get_percentage_difference(burst_kbps, expected_kbps) > 100
 
-    # The second time we use iperf to send bytes we need to see that the burst
-    # was consumed and that the transmit rate is now equal to the rate limit.
-    # We are sending the amount of bytes that can be sent in 1 sec with rate
-    # limiting enabled.
-    iperf_cmd = '{} -c {} -n{} -f KBytes'.format(
-        IPERF_BINARY,
-        host_ips[2],
-        rate_limit_bps
-    )
-    iperf_out = _run_iperf_on_guest(test_microvm, iperf_cmd, guest_ips[2])
-    iperf_out_time, iperf_out_bw = _process_iperf_output(iperf_out)
-
-    # Test that the bandwidth we obtained first time is at least as two times
-    # higher than the one obtained when rate limiting is on.
-    assert _get_difference(burst_bw_first_time_achieved, iperf_out_bw) > 100
-
-    # Test that the bandwidth we obtained second time is at least two times
-    # lower than the one obtained when no rate limiting is in place.
-    assert _get_difference(rate_no_limit_bytes, iperf_out_bw) > 100
-
-    burst_consumed_time_achieved = iperf_out_time
-    # We expect it to take around 1 sec now.
-    burst_consumed_time_expected = 1
-
-    assert (
-            _get_difference(
-                burst_consumed_time_achieved,
-                burst_consumed_time_expected
-            )
-            < MAX_TIME_DIFF
-    )
+    # Since the burst should be consumed, check rate limit is in place.
+    _check_tx_bandwidth(test_microvm, guest_ips[2], host_ips[2], expected_kbps)
 
 
 def _check_rx_rate_limiting(test_microvm, guest_ips):
@@ -255,174 +214,142 @@ def _check_rx_rate_limiting(test_microvm, guest_ips):
     _start_iperf_on_guest(test_microvm, guest_ips[0])
 
     # First step: get the transfer rate when no rate limiting is enabled.
-    # We are receiving the result in KBytes from iperf; 1000 converts to Bytes.
-    iperf_cmd = '{} {} -c {} -t{} -f KBytes -O {}'.format(
+    # We are receiving the result in KBytes from iperf.
+    iperf_cmd = '{} {} -c {} -t {} -f KBytes -O {} -w {} -N'.format(
         test_microvm.jailer.netns_cmd_prefix(),
         IPERF_BINARY,
         guest_ips[0],
         IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME
+        IPERF_OMIT_TIME,
+        IPERF_TCP_WINDOW
     )
+    print("Run guest RX iperf with no rate-limit")
     iperf_out = _run_local_iperf(iperf_cmd)
-    iperf_out = _process_iperf_output(iperf_out)[1]
-
-    rate_no_limit_bytes = float(iperf_out)
-
-    # Second step: get the number of bytes when rate limiting is on.
+    print(iperf_out)
+    _, rate_no_limit_kbps = _process_iperf_output(iperf_out)
+    print("RX rate_no_limit_kbps: {}".format(rate_no_limit_kbps))
 
     # Calculate the number of bytes that are expected to be sent
     # in each second once the rate limiting is enabled.
-    rate_limit_bps = 1000 * RATE_LIMIT_BYTES / float(RATE_LIMIT_REFILL_TIME)
+    expected_kbps = int(RATE_LIMIT_BYTES / (REFILL_TIME_MS / 1000.0) / 1024)
+    print("Configured RX expected_kbps: {}".format(expected_kbps))
 
-    # Use iperf for 3 seconds to get the number of bytes it sent with rate
-    # limiting on.
-    iperf_cmd = '{} {} -c {} -t{} -f KBytes'.format(
-        test_microvm.jailer.netns_cmd_prefix(),
-        IPERF_BINARY,
-        guest_ips[1],
-        IPERF_TRANSMIT_TIME
-    )
-    iperf_out = _run_local_iperf(iperf_cmd)
-    iperf_out = _process_iperf_output(iperf_out)[1]
-    rate_limit_bytes_achieved = 1000 * float(iperf_out)
-    rate_limit_bytes_expected = rate_limit_bps
+    # Sanity check that bandwidth with no rate limiting is at least double
+    # than the one expected when rate limiting is in place.
+    assert _get_percentage_difference(rate_no_limit_kbps, expected_kbps) > 100
 
-    # Assert on the bytes expected and achieved with rate limiting on; we are
-    # expecting a difference no bigger than MAX_RATE_LIMIT_BYTES_DIFF
-    assert (
-            _get_difference(
-                rate_limit_bytes_achieved,
-                rate_limit_bytes_expected
-            )
-            < MAX_BYTES_DIFF_PERCENTAGE
-    )
+    # Second step: check bandwith when rate limiting is on.
+    _check_rx_bandwidth(test_microvm, guest_ips[1], expected_kbps)
 
     # Third step: get the number of bytes when rate limiting is on and there is
     # an initial burst size from where to consume.
 
-    # Use iperf to obtain the time interval that a BURST_SIZE (way larger
-    # than the bucket's size) can be sent over the network.
-    iperf_cmd = '{} {} -c {} -n{} -f KBytes'.format(
+    print("Run guest TX iperf with exact burst size")
+    # Use iperf to obtain the bandwidth when there is burst to consume from.
+    iperf_cmd = '{} {} -c {} -n {} -f KBytes -w {} -N'.format(
         test_microvm.jailer.netns_cmd_prefix(),
         IPERF_BINARY,
         guest_ips[2],
-        BURST_SIZE
+        BURST_SIZE,
+        IPERF_TCP_WINDOW
     )
     iperf_out = _run_local_iperf(iperf_cmd)
+    print(iperf_out)
+    _, burst_kbps = _process_iperf_output(iperf_out)
+    print("RX burst_kbps: {}".format(burst_kbps))
+    # Test that the burst bandwidth is at least as two times the rate limit.
+    assert _get_percentage_difference(burst_kbps, expected_kbps) > 100
 
-    # iperf will give variable number of output lines depending on how much
-    # time it took to send the amount specified.
-    burst_bw_first_time_achieved = _process_iperf_output(iperf_out)[1]
-
-    # The second time we use iperf to send bytes we need to see that the burst
-    # was consumed and that the transmit rate is now equal to the rate limit.
-    # We are sending the amount of bytes that can be sent in 1 sec with rate
-    # rate limiting enabled.
-    iperf_cmd = '{} {} -c {} -n{} -f KBytes'.format(
-        test_microvm.jailer.netns_cmd_prefix(),
-        IPERF_BINARY,
-        guest_ips[2],
-        rate_limit_bps
-    )
-    iperf_out = _run_local_iperf(iperf_cmd)
-    iperf_out_time, iperf_out_bw = _process_iperf_output(iperf_out)
-
-    # Test that the bandwidth we obtained first time is at least two times
-    # higher than the one obtained when rate limiting is on.
-    assert _get_difference(burst_bw_first_time_achieved, iperf_out_bw) > 100
-    # Test that the bandwidth we obtained second time is at least two times
-    # lower than the one obtained when no rate limiting is in place.
-    assert _get_difference(rate_no_limit_bytes, iperf_out_bw) > 100
-
-    burst_consumed_time_achieved = iperf_out_time
-    # We expect it to take around 1 sec now.
-    burst_consumed_time_expected = 1
-
-    assert (
-            _get_difference(
-                burst_consumed_time_achieved,
-                burst_consumed_time_expected
-            )
-            < MAX_TIME_DIFF
-    )
+    # Since the burst should be consumed, check rate limit is in place.
+    _check_rx_bandwidth(test_microvm, guest_ips[2], expected_kbps)
 
 
 def _check_tx_rate_limit_patch(test_microvm, guest_ips, host_ips):
     """Patch the TX rate limiters and check the new limits."""
-    bucket_size = int(RATE_LIMIT_BYTES / 2)
-    bandwidth_kb = int(bucket_size / (RATE_LIMIT_REFILL_TIME/1000.0) / 1024)
+    bucket_size = int(RATE_LIMIT_BYTES * 2)
+    expected_kbps = int(bucket_size / (REFILL_TIME_MS / 1000.0) / 1024)
 
     # Check that a TX rate limiter can be applied to a previously unlimited
     # interface.
     _patch_iface_bw(test_microvm, "1", "TX", bucket_size)
-    _check_tx_bandwidth(test_microvm, guest_ips[0], host_ips[0], bandwidth_kb)
+    _check_tx_bandwidth(test_microvm, guest_ips[0], host_ips[0], expected_kbps)
 
     # Check that a TX rate limiter can be updated.
     _patch_iface_bw(test_microvm, "2", "TX", bucket_size)
-    _check_tx_bandwidth(test_microvm, guest_ips[1], host_ips[1], bandwidth_kb)
+    _check_tx_bandwidth(test_microvm, guest_ips[1], host_ips[1], expected_kbps)
 
 
 def _check_rx_rate_limit_patch(test_microvm, guest_ips):
     """Patch the RX rate limiters and check the new limits."""
-    bucket_size = int(RATE_LIMIT_BYTES / 2)
-    bandwidth_kb = int(bucket_size / (RATE_LIMIT_REFILL_TIME/1000.0) / 1024)
+    bucket_size = int(RATE_LIMIT_BYTES * 2)
+    rate_limit_kbps = int(bucket_size / (REFILL_TIME_MS / 1000.0) / 1024)
 
     # Check that an RX rate limiter can be applied to a previously unlimited
     # interface.
     _patch_iface_bw(test_microvm, "1", "RX", bucket_size)
-    _check_rx_bandwidth(test_microvm, guest_ips[0], bandwidth_kb)
+    _check_rx_bandwidth(test_microvm, guest_ips[0], rate_limit_kbps)
 
     # Check that an RX rate limiter can be updated.
     _patch_iface_bw(test_microvm, "2", "RX", bucket_size)
-    _check_rx_bandwidth(test_microvm, guest_ips[1], bandwidth_kb)
+    _check_rx_bandwidth(test_microvm, guest_ips[1], rate_limit_kbps)
 
 
 def _check_tx_bandwidth(
         test_microvm,
         guest_ip,
         host_ip,
-        expected_bw_kb
+        expected_kbps
 ):
     """Check that the rate-limited TX bandwidth is close to what we expect.
 
     At this point, a daemonized iperf3 server is expected to be running on
     the host.
     """
-    iperf_cmd = "{} -c {} -t {} -f KBytes -O {}".format(
+    print("Check guest TX rate-limit; expected kbps {}".format(expected_kbps))
+    iperf_cmd = '{} -c {} -t {} -f KBytes -O {} -w {} -N'.format(
         IPERF_BINARY,
         host_ip,
         IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME
+        IPERF_OMIT_TIME,
+        IPERF_TCP_WINDOW
     )
 
     iperf_out = _run_iperf_on_guest(test_microvm, iperf_cmd, guest_ip)
-    _, observed_bw = _process_iperf_output(iperf_out)
+    print(iperf_out)
+    _, observed_kbps = _process_iperf_output(iperf_out)
+    print("TX observed_kbps: {}".format(observed_kbps))
 
-    diff_pc = _get_difference(observed_bw, expected_bw_kb)
+    diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
+    print("TX calculated diff percentage: {}\n".format(diff_pc))
     assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
 
 
 def _check_rx_bandwidth(
         test_microvm,
         guest_ip,
-        expected_bw_kb
+        expected_kbps
 ):
     """Check that the rate-limited RX bandwidth is close to what we expect.
 
     At this point, a daemonized iperf3 server is expected to be running on
     the guest.
     """
-    iperf_cmd = "{} {} -c {} -t {} -f KBytes -O {}".format(
+    print("Check guest RX rate-limit; expected kbps {}".format(expected_kbps))
+    iperf_cmd = "{} {} -c {} -t {} -f KBytes -O {} -w {} -N".format(
         test_microvm.jailer.netns_cmd_prefix(),
         IPERF_BINARY,
         guest_ip,
         IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME
+        IPERF_OMIT_TIME,
+        IPERF_TCP_WINDOW
     )
     iperf_out = _run_local_iperf(iperf_cmd)
-    _, observed_bw = _process_iperf_output(iperf_out)
+    print(iperf_out)
+    _, observed_kbps = _process_iperf_output(iperf_out)
+    print("RX observed_kbps: {}".format(observed_kbps))
 
-    diff_pc = _get_difference(observed_bw, expected_bw_kb)
+    diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
+    print("RX calculated diff percentage: {}\n".format(diff_pc))
     assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
 
 
@@ -438,7 +365,7 @@ def _patch_iface_bw(test_microvm, iface_id, rx_or_tx, new_bucket_size):
         "{}_rate_limiter".format(rx_or_tx.lower()): {
             'bandwidth': {
                 'size': new_bucket_size,
-                'refill_time': RATE_LIMIT_REFILL_TIME
+                'refill_time': REFILL_TIME_MS
             }
         }
     }
@@ -455,7 +382,7 @@ def _start_iperf_on_guest(test_microvm, hostname):
     ssh_connection.execute_command(iperf_cmd)
 
     # Wait for the iperf daemon to start.
-    time.sleep(2)
+    time.sleep(1)
 
 
 def _run_iperf_on_guest(test_microvm, iperf_cmd, hostname):
@@ -482,7 +409,7 @@ def _start_local_iperf(netns_cmd_prefix):
     utils.run_cmd(iperf_cmd)
 
     # Wait for the iperf daemon to start.
-    time.sleep(2)
+    time.sleep(1)
 
 
 def _run_local_iperf(iperf_cmd):
@@ -491,34 +418,32 @@ def _run_local_iperf(iperf_cmd):
     return process.stdout
 
 
-def _get_difference(current, previous):
+def _get_percentage_difference(measured, base):
     """Return the percentage delta between the arguments."""
-    if current == previous:
+    if measured == base:
         return 0
     try:
-        return (abs(current - previous) / previous) * 100.0
+        return (abs(measured - base) / base) * 100.0
     except ZeroDivisionError:
-        # It means previous and only previous is 0.
+        # It means base and only base is 0.
         return 100.0
 
 
+def _process_iperf_line(line):
+    """Parse iperf3 summary line and return test time and bandwidth."""
+    time = line.split('  ')[2].split('-')[1].strip().split(" ")[0]
+    bw = line.split('  ')[5].split(' ')[0].strip()
+    return float(time), float(bw)
+
+
 def _process_iperf_output(iperf_out):
-    """Parse iperf3 output and return test time and bandwidth."""
-    found_line = 0
+    """Parse iperf3 output and return average test time and bandwidth."""
     iperf_out_lines = iperf_out.splitlines()
     for line in iperf_out_lines:
-        if line.find('- - - - - - - -') != -1:
-            found_line += 1
-
-        if found_line == 3:
-            iperf_out_time = line.split('  ')[2].split(
-                '-'
-            )[1].strip().split(" ")[0]
-            iperf_out_bw = line.split('  ')[5].split(
-                ' '
-            )[0].strip()
-            break
-        if found_line > 0:
-            # Skip the first 3 lines after the first line containing `------`
-            found_line += 1
+        if line.find('sender') != -1:
+            send_time, send_bw = _process_iperf_line(line)
+        if line.find('receiver') != -1:
+            rcv_time, rcv_bw = _process_iperf_line(line)
+    iperf_out_time = (send_time + rcv_time) / 2.0
+    iperf_out_bw = (send_bw + rcv_bw) / 2.0
     return float(iperf_out_time), float(iperf_out_bw)

--- a/tests/integration_tests/functional/test_rate_limiter.py
+++ b/tests/integration_tests/functional/test_rate_limiter.py
@@ -12,10 +12,6 @@ IPERF_BINARY = 'iperf3'
 # Interval used by iperf to get maximum bandwidth
 IPERF_TRANSMIT_TIME = 4
 
-# Interval to omit statistics collection at the beginning of the test, to skip
-# past the TCP slow-start period
-IPERF_OMIT_TIME = 1
-
 # Use a fixed-size TCP window so we get constant flow
 IPERF_TCP_WINDOW = '1000K'
 
@@ -162,11 +158,10 @@ def _check_tx_rate_limiting(test_microvm, guest_ips, host_ips):
 
     # First step: get the transfer rate when no rate limiting is enabled.
     # We are receiving the result in KBytes from iperf.
-    iperf_cmd = '{} -c {} -t {} -f KBytes -O {} -w {} -N'.format(
+    iperf_cmd = '{} -c {} -t {} -f KBytes -w {} -N'.format(
         IPERF_BINARY,
         host_ips[0],
         IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME,
         IPERF_TCP_WINDOW
     )
     print("Run guest TX iperf with no rate-limit")
@@ -215,12 +210,11 @@ def _check_rx_rate_limiting(test_microvm, guest_ips):
 
     # First step: get the transfer rate when no rate limiting is enabled.
     # We are receiving the result in KBytes from iperf.
-    iperf_cmd = '{} {} -c {} -t {} -f KBytes -O {} -w {} -N'.format(
+    iperf_cmd = '{} {} -c {} -t {} -f KBytes -w {} -N'.format(
         test_microvm.jailer.netns_cmd_prefix(),
         IPERF_BINARY,
         guest_ips[0],
         IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME,
         IPERF_TCP_WINDOW
     )
     print("Run guest RX iperf with no rate-limit")
@@ -306,22 +300,51 @@ def _check_tx_bandwidth(
     the host.
     """
     print("Check guest TX rate-limit; expected kbps {}".format(expected_kbps))
-    iperf_cmd = '{} -c {} -t {} -f KBytes -O {} -w {} -N'.format(
+    observed_kbps = _get_tx_bandwidth_with_duration(
+        test_microvm,
+        guest_ip,
+        host_ip,
+        IPERF_TRANSMIT_TIME
+    )
+
+    diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
+    print("TX calculated diff percentage: {}\n".format(diff_pc))
+
+    if diff_pc >= MAX_BYTES_DIFF_PERCENTAGE:
+        print("Short duration test failed. Try another run with 10x duration.")
+
+        observed_kbps = _get_tx_bandwidth_with_duration(
+            test_microvm,
+            guest_ip,
+            host_ip,
+            10 * IPERF_TRANSMIT_TIME
+        )
+        diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
+        print("TX calculated diff percentage: {}\n".format(diff_pc))
+
+        assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
+
+
+def _get_tx_bandwidth_with_duration(
+        test_microvm,
+        guest_ip,
+        host_ip,
+        duration
+):
+    """Check that the rate-limited TX bandwidth is close to what we expect."""
+    iperf_cmd = '{} -c {} -t {} -f KBytes -w {} -N'.format(
         IPERF_BINARY,
         host_ip,
-        IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME,
+        duration,
         IPERF_TCP_WINDOW
     )
 
     iperf_out = _run_iperf_on_guest(test_microvm, iperf_cmd, guest_ip)
     print(iperf_out)
+
     _, observed_kbps = _process_iperf_output(iperf_out)
     print("TX observed_kbps: {}".format(observed_kbps))
-
-    diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
-    print("TX calculated diff percentage: {}\n".format(diff_pc))
-    assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
+    return observed_kbps
 
 
 def _check_rx_bandwidth(
@@ -335,22 +358,48 @@ def _check_rx_bandwidth(
     the guest.
     """
     print("Check guest RX rate-limit; expected kbps {}".format(expected_kbps))
-    iperf_cmd = "{} {} -c {} -t {} -f KBytes -O {} -w {} -N".format(
+    observed_kbps = _get_rx_bandwidth_with_duration(
+        test_microvm,
+        guest_ip,
+        IPERF_TRANSMIT_TIME
+    )
+
+    diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
+    print("RX calculated diff percentage: {}\n".format(diff_pc))
+
+    if diff_pc >= MAX_BYTES_DIFF_PERCENTAGE:
+        print("Short duration test failed. Try another run with 10x duration.")
+
+        observed_kbps = _get_rx_bandwidth_with_duration(
+            test_microvm,
+            guest_ip,
+            10 * IPERF_TRANSMIT_TIME
+        )
+        diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
+        print("TX calculated diff percentage: {}\n".format(diff_pc))
+
+        assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
+
+
+def _get_rx_bandwidth_with_duration(
+        test_microvm,
+        guest_ip,
+        duration
+):
+    """Check that the rate-limited RX bandwidth is close to what we expect."""
+    iperf_cmd = "{} {} -c {} -t {} -f KBytes -w {} -N".format(
         test_microvm.jailer.netns_cmd_prefix(),
         IPERF_BINARY,
         guest_ip,
-        IPERF_TRANSMIT_TIME,
-        IPERF_OMIT_TIME,
+        duration,
         IPERF_TCP_WINDOW
     )
     iperf_out = _run_local_iperf(iperf_cmd)
     print(iperf_out)
+
     _, observed_kbps = _process_iperf_output(iperf_out)
     print("RX observed_kbps: {}".format(observed_kbps))
-
-    diff_pc = _get_percentage_difference(observed_kbps, expected_kbps)
-    print("RX calculated diff percentage: {}\n".format(diff_pc))
-    assert diff_pc < MAX_BYTES_DIFF_PERCENTAGE
+    return observed_kbps
 
 
 def _patch_iface_bw(test_microvm, iface_id, rx_or_tx, new_bucket_size):
@@ -431,9 +480,9 @@ def _get_percentage_difference(measured, base):
 
 def _process_iperf_line(line):
     """Parse iperf3 summary line and return test time and bandwidth."""
-    time = line.split('  ')[2].split('-')[1].strip().split(" ")[0]
-    bw = line.split('  ')[5].split(' ')[0].strip()
-    return float(time), float(bw)
+    test_time = line.split('  ')[2].split('-')[1].strip().split(" ")[0]
+    test_bw = line.split('  ')[5].split(' ')[0].strip()
+    return float(test_time), float(test_bw)
 
 
 def _process_iperf_output(iperf_out):

